panther_analysis_tool/log_schemas/user_defined.py

```diff
@@ -0,0 +1,279 @@
+                    'isManaged': False
+                }
+            )
+        )
+
+    def put_schema(self, name: str, definition: str, revision: int,  # pylint: disable=too-many-arguments
+                   description: str, reference_url: str) -> Tuple[bool, dict]:
+        """
+        Update a custom schema
+        """
+        return self._invoke(
+            self._create_lambda_request(
+                endpoint=self._PUT_SCHEMA_ENDPOINT,
+                payload=dict(
+                    name=name,
+                    referenceURL=reference_url,
+                    description=description,
+                    spec=definition,
+                    revision=revision,
+                )
+            )
+        )
+
+    def _invoke(self, request: dict) -> Tuple[bool, dict]:
+        response = self.lambda_client.invoke(**request)
+        response = json.loads(response["Payload"].read().decode("utf-8"))
+        api_error = response.get("error")
+        if api_error is not None:
+            return False, response
+        return True, response
+
+    def _create_lambda_request(self, endpoint: str, payload: dict) -> dict:
+        return dict(
+            FunctionName=self._LAMBDA_NAME,
+            InvocationType="RequestResponse",
+            Payload=json.dumps({endpoint: payload})
+        )
+
+
+@dataclass
+class UploaderResult:
+    filename: str
+    name: Optional[str]
+    api_response: Optional[Dict[str, Any]] = None
+    definition: Optional[Dict[str, Any]] = None
+    error: Optional[str] = None
+    existed: Optional[bool] = None
+
+
+@dataclass
+class ProcessedFile:
+    error: Optional[str] = None
+    raw: str = ''
+    yaml: Optional[Dict[str, Any]] = None
+
+
+class Uploader:
+    _SCHEMA_NAME_PREFIX = 'Custom.'
+
+    def __init__(self, path: str):
+        self._path = path
+        self._files: Optional[List[str]] = None
+        self._api_client: Optional[Client] = None
+        self._existing_schemas: Optional[List[Dict[str, Any]]] = None
+
+    @property
+    def api_client(self) -> Client:
+        if self._api_client is None:
+            self._api_client = Client()
+        return self._api_client
+
+    @property
+    def files(self) -> List[str]:
+        if self._files is None:
+            self._files = discover_files(self._path)
+        return self._files
+
+    @property
+    def existing_schemas(self) -> List[Dict[str, Any]]:
+        if self._existing_schemas is None:
+            success, response = self.api_client.list_schemas()
+            if not success:
+                raise RuntimeError('unable to retrieve custom schemas')
+            self._existing_schemas = response['results']
+        return self._existing_schemas
+
+    def find_schema(self, name: str) -> Optional[Dict[str, Any]]:
+        for schema in self.existing_schemas:
+            if schema['name'] == name:
+                return schema
+        return None
+
+    def process(self) -> List[UploaderResult]:
```

Comment:
Could you add description to this process? I am particularly concerned about the logic that first fetches schemas before uploading them. It will not be obvious to everyone why this has to happen.

Reply:
Sure, I will write some notes on the matching revision requirement on behalf of the API endpoint.